Modules/posixmodule.c

```diff
@@ -1652,56 +1589,59 @@ win32_xstat_impl(const wchar_t *path, struct _Py_stat_struct *result,
+            lastError != ERROR_SHARING_VIOLATION &&
+            lastError != ERROR_INVALID_PARAMETER)
             return -1;
         /* Could not get attributes on open file. Fall back to
            reading the directory. */
         if (!attributes_from_dir(path, &info, &reparse_tag))
             /* Very strange. This should not fail now */
             return -1;
+        is_link = _Py_is_reparse_link(path, reparse_tag);
         if (info.dwFileAttributes & FILE_ATTRIBUTE_REPARSE_POINT) {
-            if (traverse) {
+            if (!is_link || traverse) {
                 /* Should traverse, but could not open reparse point handle */
                 SetLastError(lastError);
                 return -1;
             }
         }
     } else {
-        if (!GetFileInformationByHandle(hFile, &info)) {
+        // Get file attributes + reparse tag first
+        if (!GetFileInformationByHandleEx(hFile, FileAttributeTagInfo,
+                                          &taginfo, sizeof(taginfo))) {
             CloseHandle(hFile);
             return -1;
         }
-        if (info.dwFileAttributes & FILE_ATTRIBUTE_REPARSE_POINT) {
-            if (!win32_get_reparse_tag(hFile, &reparse_tag))
-                return -1;
-
-            /* Close the outer open file handle now that we're about to
-               reopen it with different flags. */
-            if (!CloseHandle(hFile))
-                return -1;
+        if (taginfo.FileAttributes & FILE_ATTRIBUTE_REPARSE_POINT) {
+            is_link = _Py_is_reparse_link(path, taginfo.ReparseTag);
 
-            if (traverse) {
-                /* In order to call GetFinalPathNameByHandle we need to open
-                   the file without the reparse handling flag set. */
-                hFile2 = CreateFileW(
-                           path, FILE_READ_ATTRIBUTES, FILE_SHARE_READ,
-                           NULL, OPEN_EXISTING,
-                           FILE_ATTRIBUTE_NORMAL|FILE_FLAG_BACKUP_SEMANTICS,
-                           NULL);
-                if (hFile2 == INVALID_HANDLE_VALUE)
+            if (!is_link || traverse) {
+                /* Close the outer open file handle now that we're about to
+                   reopen it with different flags. */
+                if (!CloseHandle(hFile))
                     return -1;
-
-                if (!get_target_path(hFile2, &target_path))
+                // FILE_FLAG_OPEN_REPARSE_POINT to follow reparses:
+                hFile = CreateFileW(
+                    path, FILE_READ_ATTRIBUTES, 0,
+                    NULL, OPEN_EXISTING,
+                    FILE_ATTRIBUTE_NORMAL|FILE_FLAG_BACKUP_SEMANTICS,
+                    NULL);
+                if (hFile == INVALID_HANDLE_VALUE)
                     return -1;
-
-                code = win32_xstat_impl(target_path, result, FALSE);
-                PyMem_RawFree(target_path);
-                return code;
+                if (!GetFileInformationByHandle(hFile, &info)) {
```

Comment:
Why do this here instead of letting it fall through to a common `GetFileInformationByHandle` call? `hFile` was closed and re-opened to allow a common call.